### PR TITLE
Setup goreleaser for gwctl to generate binaries for various platforms

### PR DESCRIPTION
**Pull request by gauravkghildiyal**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
- Introduce `gwctl version` command to print binary information.
  ```
  ❯ gwctl version
  Version: 1.0.0-SNAPSHOT-ca344b9a
  Git Commit: ca344b9aa43ad5acc81235b25097fd20d8e97068
  Build Date: 2024-05-01T06:29:53Z
  ```
- Setup goreleaser to build binaries for various platforms

NOTES for the reviewer:
- LICENSE file is added to the gwctl subdirectory so that it get's included in the archives generated by goreleaser.
- Actual steps of how maintainers will make use of goreleaser to generate the binaries will be added independently (likely after some discussion)

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: gwctl/.goreleaser.yaml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+# For local testing, use:
+# 
+#   goreleaser release --snapshot --clean
+project_name: gwctl
+
+before:
+  hooks:
+    # Ensure the project builds successfully before release.
+    - go mod tidy
+    - go build -o ./dist/gwctl ./main.go
```
Comment (youngnick) [inline gwctl/.goreleaser.yaml:12]: At some point, both this and the repo CI should probably include a test suite. Much larger change, non-blocking for this one, but something to put on your todo.

Comment (gauravkghildiyal) [reply gwctl/.goreleaser.yaml:12]: Absolutely yes.

File: gwctl/.goreleaser.yaml [hunk 2]
```diff
@@ -0,0 +1,35 @@
+# https://goreleaser.com/
+#
+# For local testing, use:
+# 
+#   goreleaser release --snapshot --clean
+project_name: gwctl
```
Comment (pottekkat) [inline gwctl/.goreleaser.yaml:6]: Do we intend to keep gwctl as a standalone binary or as a kubectl plugin? If it is both, then wouldn't it be better to name the binary `kubectl-*`?

File: gwctl/cmd/root.go [hunk 3]
```diff
@@ -94,3 +96,15 @@ func getParams(path string) *cmdutils.CmdParams {
...
+	cmd := &cobra.Command{
+		Use:   "version",
+		Short: "Print the version information of gwctl",
+		Long:  `Print the version information of gwctl, including version, git commit and build date.`,
+		Run: func(*cobra.Command, []string) {
+			fmt.Println(version.GetVersionInfo())
+		},
+	}
+	return cmd
+}
```
Comment (pottekkat) [inline gwctl/cmd/root.go:110]: nit: Shouldn't this be moved to a separate file like the rest of the subcommands?

**Comment by mlavacca** [comment]
I'll take a look into it

/cc

File: gwctl/.goreleaser.yaml [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    goarch:
+      - 386
+      - amd64
+      - arm64
+    ldflags:
+      # Reduce the size of the binary. Ref: https://pkg.go.dev/cmd/link
+      - -s -w 
+      # Embed versioning and build information in the binary.
+      - -X sigs.k8s.io/gateway-api/gwctl/pkg/version.version={{.Version}}
+      - -X sigs.k8s.io/gateway-api/gwctl/pkg/version.gitCommit={{.Commit}}
```
Comment (guicassolato) [inline gwctl/.goreleaser.yaml:34]: I think `.Commit` is [deprecated](https://goreleaser.com/customization/templates/#common-fields). Should we use `.FullCommit` perhaps instead?

```suggestion
      - -X sigs.k8s.io/gateway-api/gwctl/pkg/version.gitCommit={{.FullCommit}}
```

Comment (gauravkghildiyal) [reply gwctl/.goreleaser.yaml:34]: Ah thanks! Didn't know about the deprecation. Will update.

File: gwctl/.goreleaser.yaml (hunk 2 above)
Comment (gauravkghildiyal) [reply gwctl/.goreleaser.yaml:6]: This PR only intended to make this available as a standalone binary, but you are right for kubectl plugin that is indeed a better name. (I'll attempt at addressing this in the new PR)

File: gwctl/cmd/root.go (hunk 3 above)
Comment (gauravkghildiyal) [reply gwctl/cmd/root.go:110]: Sure can! Thanks

**Comment by gauravkghildiyal** [comment]
Thanks for the reviews here everyone! Closing this in favor of resolving this within https://github.com/kubernetes-sigs/gwctl. 

/close

**Pull request closed without merging**